hnn_core/network.py

```diff
@@ -228,42 +219,106 @@ def _add_external_feeds(self):
         # Global number of external inputs ... automatic counting
         # makes more sense
         # p_unique represent ext inputs that are going to go to each cell
-        self.p_common, self.p_unique = create_pext(self.params,
-                                                   self.params['tstop'])
-        self.n_common_feeds = len(self.p_common)
+        self._p_common, self._p_unique = create_pext(self.params,
+                                                     self.params['tstop'])
+        self._n_common_feeds = len(self._p_common)
 
         # 'position' the artificial cells arbitrarily in the origin of the
         # network grid. Non-biophysical cell placement is irrelevant
+        # However, they must be present to be included in gid_ranges!
         origin = self.pos_dict['origin']
 
         # COMMON FEEDS
-        self.pos_dict['common'] = [origin for i in range(self.n_common_feeds)]
+        self.pos_dict['common'] = [origin for i in range(self._n_common_feeds)]
 
         # UNIQUE FEEDS
-        for key in self.p_unique.keys():
+        for key in self._p_unique.keys():
             # create the pos_dict for all the sources
             self.pos_dict[key] = [origin for i in range(self.n_cells)]
 
         # Now add the names of the feeds to a list
         self.feedname_list.append('common')
         # grab the keys for the unique set of inputs and sort the names
         # append them to the src list along with the number of cells
-        unique_keys = sorted(self.p_unique.keys())
+        unique_keys = sorted(self._p_unique.keys())
         self.feedname_list += unique_keys
 
-    def _create_gid_dict(self):
-        """Creates gid dicts."""
+        # Every time pos_dict is updated, gid_ranges must be updated too
+        self._update_gid_ranges()
+
+        # Create the feed dynamics (event_times)
+        self._instantiate_feeds()
+
+    def _instantiate_feeds(self):
+        '''Creates event_time vectors for all feeds and all trials
+
+        NB this must be a separate method because dipole.py:simulate_dipole
+        accepts an n_trials-argument, which overrides the N_trials-parameter
+        used at intialisation time. The good news is that only the event_times
+        need to be recalculated, all the GIDs etc remain the same.
+        '''
+        # each trial needs unique event time vectors
+
+        self.trial_event_times = []  # reset if called again from dipole.py
+        n_trials = self.params['N_trials']
+
+        cur_params = self.params.copy()  # these get mangled below!
+        for trial_idx in range(n_trials):
+
+            prng_seedcore_initial = cur_params['prng_*'].copy()
+            for param_key in prng_seedcore_initial.keys():
+                cur_params[param_key] =\
+                    prng_seedcore_initial[param_key] + trial_idx
```

Comment:
we should rethink this approach now that a lot of `feed.py` code is simplified in #204 and we begin to understand what that code is doing. I tagged you in #113

Reply:
Why not merge #204 first, then rebase and fix this (more invasive) PR on that? I also predict  #113 will then just melt away :)